docs/yul.rst

```diff
@@ -976,6 +976,47 @@ within one Yul subobject. If at least one ``memoryguard`` call is found in a sub
 the additional optimiser steps will be run on it.
 
 
+verbatim
+^^^^^^^^
+
+The set of ``verbatim`` builtin functions lets you create bytecode for opcodes
+that are not known to the Yul compiler. It also allows you to create
+bytecode sequences that will not be touched by the optimizer.
+
+The functions are ``verbatim_<n>i_<m>o("<data>", ...)``, where
+ - ``n`` is a decimal that specifies the number of input stack slots / variables
+ - ``m`` is a decimal that specifies the number of output stack slots / variables
+ - ``data`` is a string literal that contain the sequence of bytes
+
+If ``n`` is or ``m`` is zero, the full part can be left out.
+For example, ``verbatim_2o`` is a function that outputs two stack slots
+and does not consume any, while ``verbatim`` has no effect on the stack
+and ``verbatim_1i`` consumes one stack slot.
+
+If you for example want to define a function that multiplies the input
+by two, without the optimizer touching the constant two, you can use
+
+.. code-block::
+
+    let x := ...
+    let double := verbatim_1i_1o(hex"600202, x)
+
+This code will result in a ``dupN`` opcode to retrieve ``x``
+directly followed by ``600202``. The code is assumed to
+consume the copied value of ``x`` and produce the result
+on the top of the stack. The compiler then generates code
+to allocate a stack slot for ``double`` and store the result there.
+
+Since you could be using unknown opcodes, the ``verbatim`` builtins
+are being treated as having worst-case side-effects, i.e. the optimizer
+assumes they can potentially modify all state. On the other hand,
+it assumes that such builtins will leave the stack unchanged, except for the
+input and output values.
+
+The optimizer works if there are jumps inside the ``verbatim`` code,
+but control-flow should not jump out of or into this code.
```

Comment:
Yeah, ok, @hrkrshnn is definitely right in that there's more things to consider here. A few I remember are:
- Accessing (even reading) stack contents outside the designated input variables results in undefined behaviour.
- Violating the specified stack balance results in undefined behaviour.
- Use of `gas` and `pc` is discouraged unless you know exactly what you're doing.

And maybe even some technical details like
"The optimizer assumes that `verbatim` can read and/or modify the entire state (i.e. storage and memory) and treats it accordingly. However, `verbatim` may not rely on the surrounding code (or the surrounding stack layout) be unmodified by the optimizer or ...." [probably some more stuff]

Depending on how we proceed with stack limit evasion we may also have to specify that yul code containing a ``memoryguard`` will need to uphold its memory guarantees (even though that's actually implicit in how the memoryguard works anyways).

Reply:
Maybe we should even specify that we take the liberty to define new restrictions on verbatim in non-breaking releases or sth like that, just as a precaution against this preventing us from doing stuff because of a weird "this is breaking because of verbatim" notion...

Reply:
I expanded this section a little. I did not add the point about `gas` and `pc` - is this special about verbatim? Note that using `pc` is the only way to have jumps inside the verbatim bytecode.